src/sentry/search/snuba/backend.py

```diff
@@ -534,25 +534,37 @@ def _get_queryset_conditions(
             queryset_conditions.update(
                 {"issue.type": QCallbackCondition(lambda types: Q(type__in=types))}
             )
-            queryset_conditions.update(
-                {
-                    "message": QCallbackCondition(
-                        lambda query: Q(type=GroupType.ERROR.value)
-                        | Q(
-                            type__in=(
-                                GroupType.PERFORMANCE_N_PLUS_ONE.value,
-                                GroupType.PERFORMANCE_SLOW_SPAN.value,
-                                GroupType.PERFORMANCE_SEQUENTIAL_SLOW_SPANS.value,
-                                GroupType.PERFORMANCE_LONG_TASK_SPANS.value,
-                                GroupType.PERFORMANCE_RENDER_BLOCKING_ASSET_SPAN.value,
-                                GroupType.PERFORMANCE_DUPLICATE_SPANS.value,
-                                GroupType.PERFORMANCE_N_PLUS_ONE_DB_QUERIES.value,
-                            ),
-                            message__icontains=query,
-                        )
-                    )
-                }
+            message_filter = next(
+                (sf for sf in search_filters or () if "message" == sf.key.name), None
             )
+            if message_filter:
+
+                def _perf_issue_message_condition(query: str) -> Q:
+                    return Q(
+                        type__in=(
+                            GroupType.PERFORMANCE_N_PLUS_ONE.value,
+                            GroupType.PERFORMANCE_SLOW_SPAN.value,
+                            GroupType.PERFORMANCE_SEQUENTIAL_SLOW_SPANS.value,
+                            GroupType.PERFORMANCE_LONG_TASK_SPANS.value,
+                            GroupType.PERFORMANCE_RENDER_BLOCKING_ASSET_SPAN.value,
+                            GroupType.PERFORMANCE_DUPLICATE_SPANS.value,
+                            GroupType.PERFORMANCE_N_PLUS_ONE_DB_QUERIES.value,
+                        ),
+                        message__icontains=query,
+                    )
+
+                queryset_conditions.update(
+                    {
+                        "message": QCallbackCondition(
+                            lambda query: Q(type=GroupType.ERROR.value)
+                            | _perf_issue_message_condition(query)
+                        )
+                        if not message_filter.is_negation
+                        # if we're negating the message search term: '!message:something' we shouldn't apply
+                        # the `type=GroupType.ERROR` check since we'll end up excluding all error issues
+                        else QCallbackCondition(lambda query: _perf_issue_message_condition(query))
```

Comment:
does this need to have the `~` negation like in the executor? also `nit` but this logic is in the reverse order from that

Reply:
The logic is a bit non-intuitive when just looking at this block of code but when negation happens all rows that satisfy the predicate gets [excluded](https://github.com/getsentry/sentry/blob/faee67a41c532cc5c897a978e56f3886e820b96e/src/sentry/search/snuba/backend.py#L300-L302).

So we don't want to exclude any rows where `type=GroupType.ERROR` since that will get handled when querying snuba.